File: src/step2point/algorithms/identity.py
from __future__ import annotations

import numpy as np

from step2point.algorithms.base import CompressionAlgorithm
from step2point.core.results import CompressionResult
from step2point.core.shower import Shower


class IdentityCompression(CompressionAlgorithm):
    """Identity compression.

    No change of the shower. Serves as an example.
    """

    name = "identity"

    def compress(self, shower: Shower) -> CompressionResult:
        return CompressionResult(
            shower=shower,
            algorithm=self.name,
            stats={
                "n_points_before": shower.n_points,
                "n_points_after": shower.n_points,
                "compression_ratio": 1.0,
                "energy_before": shower.total_energy,
                "energy_after": shower.total_energy,
            },
            debug_data={"cluster_label": np.arange(shower.n_points, dtype=np.int64)},
        )

File: cpp/src/identity.cpp
#include "step2point/algorithms.hpp"

#include <stdexcept>

namespace step2point {

CompressionResult identity(const Shower& input) {
  if (!input.valid()) {
    throw std::runtime_error("Invalid Shower passed to identity().");
  }
  CompressionResult out;
  out.shower = input;
  out.algorithm = "identity";
  out.stats.n_points_before = input.n_points();
  out.stats.n_points_after = input.n_points();
  out.stats.energy_before = input.total_energy();
  out.stats.energy_after = input.total_energy();
  out.stats.compression_ratio = 1.0;
  return out;
}

} // namespace step2point
